### ISSUE TITLE
[FEATURE] Instance Manager Resource Request Strategy

### ISSUE THREAD

**Issue by innobead** [longhorn/longhorn#6351]
## Is your feature request related to a problem? Please describe (👍 if you like this request)

The longhorn instance manager component is a mission-critical component responsible for the volume lifecycle including volume frontend, engine, and replica, so it should continue working with the right resource allocation w/o any downtime to prevent any potential resource starvation to impact its running state. (here, we are not handling involuntary issues to cause node down and then impact instance manager pods to stop working accidentally) 

Currently, a user is able to set a CPU request to it as per https://longhorn.io/docs/1.5.0/references/settings/#guaranteed-instance-manager-cpu, but it's not quite clear for users to understand why they need to have pre-allocated resources to it, even though there is no much volume usage on that node, and it turns out to a resource usage concern)

It's better to consider different strategies for users and clearly tell users how to do and what options able to choose like `no resource request (best effort)`, and `pre-requested resource (strict mode)` by considering if the node role like the node is a pure storage node w/o user workloads or only frontend w/o replicas, etc. 

Should leverage priority class for this component to ensure it's liveness, https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#priorityclass by default. 

All the above should be considered with other unclarified factors, and make a strategy global setting for it. Also, need to review the resource CPU calculation for the engine & replica when it's used by a data-intensive application, so we can use the extreme case for the estimation formula instead of a unclear theory assumption)

## Describe the solution you'd like

A clear and concise description of what you want to happen

## Additional context

cc @longhorn/dev

**Comment by derekbit** [comment]
@ejweber Please help work on this one. If you have any question, please go discuss with @PhanLe1010.

**Comment by ejweber** [comment]
We discussed this a bit at the team meeting and want to consider Kubernetes' new vertical scaling features as we design it:

- https://kubernetes.io/docs/concepts/workloads/autoscaling/#scaling-workloads-vertically
- https://kubernetes.io/docs/tasks/configure-pod-container/resize-container-resources/

**Comment by ejweber** [comment]
TLDR of the above:

InPlacePodVerticalScaling for Kubernetes pods is in Alpha in Kubernetes v1.27. It was scheduled to go Beta in v1.30, but looks like it missed that mark. It was scheduled to go GA in v1.32, but will no longer make it.

This feature could definitely be helpful for Longhorn in general, as it would allow longhorn-manager to change resource requests and limits on instance-manager pods without restarting them.

The vertical-pod-autoscalar may also be interesting, though it would be more difficult to use since (like the snapshot controller) it should be separately installed in a cluster. We could theoretically create CRs matching instance-manager pods that would dynamically adjust the CPU requests based on recent usage. Longhorn could potentially lower bound the CPU request scaling to some minimum based on our existing minimum guaranteed instance manager CPU formula, while allowing the request to grow or shrink during runtime. The user could configure the upper bound in some way to stop instance-manager from requesting all resources (and we would never allow requests to exceed 50% of the available CPU, or live migration would be impossible).

**Comment by ejweber** [comment]
We are going to target this one for v1.8.0 in conjunction with https://github.com/longhorn/longhorn/issues/8332. It is mainly because the instance-manager CPU/memory requirements cannot be readily changed under the current design. Changing them requires stopping all running engines and replicas on each node (so the instance-manager can be restarted). `InPlacePodVerticalScaling` (when it matures) should allow us to update an instance-manager resource requests/limits without restarting it, making it much more useful.

Both this ticket and https://github.com/longhorn/longhorn/issues/8332 benefit from an investigation of the Rook Cepth resource constraint model. I originally looked at https://github.com/rook/rook/blob/master/design/ceph/resource-constraints.md, but some of what is discussed there is outdated or was never implemented. For example, there is no evidence in the code of an "automatic algorithm" that determines the correct resource requests for OSDs or a `multiplier` field that scales resources without setting them explicitly. However, Rook Ceph DOES have a fairly easy to understand resource constraint management.

Rook Ceph has a single cluster-wide CRD in which default resource requests/limits can be set for all types of pods required by Ceph clusters. In addition, more specific storage cluster CRDs can also contain these resource requests and limits. There is a configuration hierarchy that ensures more specific configuration overrides less specific configuration.

https://rook.io/docs/rook/latest-release/CRDs/Cluster/ceph-cluster-crd/#cluster-wide-resources-configuration-settings

Longhorn does not currently have a good analog. We have global settings that don't necessarily lend themselves well to complex configuration (they are generally key/value pairs where the value is a string or boolean). We also have CRDs that are managed by the system (e.g. Node and InstanceManager). We may want/need a better place to store resource requests.

Rook Ceph defines minimum memory requirements for each type of pod and warns the user if they specify a value that is lower than the minimum. In addition, when a memory limit is set on a pod, Rook Ceph generally ensures that the process running inside that pod limits its memory consumption to the same amount. The current Longhorn processes cannot be restricted in that way, so memory limits may be a dangerous proposition for them.

**Comment by derekbit** [comment]
Ref: 
- https://kubernetes.io/docs/concepts/workloads/autoscaling/vertical-pod-autoscale/
- https://kubernetes.io/docs/concepts/workloads/autoscaling/vertical-pod-autoscale/#updateMode-InPlaceOrRecreate (1.35 GA)
